agent/dns.go

```diff
@@ -1169,14 +1170,27 @@ func (d *DNSServer) serviceNodeRecords(dc string, nodes structs.CheckServiceNode
 		// Add the node record
 		records := d.formatNodeRecord(node.Node, addr, qName, qType, ttl, edns, true)
 		if records != nil {
-			resp.Answer = append(resp.Answer, records...)
-			count++
-			if count == d.config.ARecordLimit {
-				// We stop only if greater than 0 or we reached the limit
-				return
+			switch records[0].(type) {
+			case *dns.CNAME:
+				// keep track of the first CNAME + associated RRs but don't add to the resp.Answer yet
+				// this will only be added if no non-CNAME RRs are found
```

Comment:
The node list gets shuffled in the serviceLookup function prior to actually formatting so it will be randomized which CNAME gets used.

This is kind of the best we can always do when mapping a single service name with fqdns for the service address.

One other approach I thought of would be to check if we knew the A/AAAA RR for the name pointed to by the CNAME and instead propagate those records up and hide the CNAME all together. This would work when the service address is a name known to Consul (either it is authoritative for it like with being a node name or Consul has recursors configured and can lookup the name that way).

However because this wouldn't work in a general case like the one that prompted this change I didn't think that was the route we should go.

Reply:
Shuffling done higher up is 👍